lib/functions/clocksource.functions.sh

```diff
@@ -0,0 +1,147 @@
+#!/bin/bash
+#==============================================================================
+# Copyright (C) 2017 Stephen F. Norledge and Alces Software Ltd.
+#
+# This file/package is part of Alces Clusterware.
+#
+# Alces Clusterware is free software: you can redistribute it and/or
+# modify it under the terms of the GNU Affero General Public License
+# as published by the Free Software Foundation, either version 3 of
+# the License, or (at your option) any later version.
+#
+# Alces Clusterware is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this package.  If not, see <http://www.gnu.org/licenses/>.
+#
+# For more information on the Alces Clusterware, please visit:
+# https://github.com/alces-software/clusterware
+#==============================================================================
+require action
+require ui
+require files
+require member
+
+cw_CLOCKSOURCE_current_file=/sys/devices/system/clocksource/clocksource0/current_clocksource
+cw_CLOCKSOURCE_config_file=${cw_ROOT}/etc/cluster-clocksource.rc
+cw_CLOCKSOURCE_sym_link=${cw_ROOT}/var/lib/alces-flight-www/customizer/cluster-clocksource.rc
+
+_clocksource_write_config() {
+    local default="$1" clocksource="$2"
+    shift 2
+
+    cat <<EOF > $cw_CLOCKSOURCE_config_file
+################################################################################
+##
+## Alces Clusterware - Clocksource configuration
+## Copyright (c) 2017 Alces Software Ltd
+##
+################################################################################
+cw_CLUSTER_default_clocksource=${default}
+cw_CLUSTER_clocksource=${clocksource}
+EOF
+}
+
+_clocksource_get_available() {
+    cat /sys/devices/system/clocksource/clocksource0/available_clocksource
+}
+
+_clocksource_validate_source() {
+    local input="$1"
+    shift
+
+    validated=$(_clocksource_get_available | grep -wo "$input")
```

Comment:
The `validated` variable can be local to this function.

Given that the script exits with an `action_die` if the proposed clock source fails validation, then there's no need for the value of `validated` to be passed around anywhere; if the script is still running after calling `_clocksource_validate_source` then you know that the proposed clock source is a valid one.